clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/questionnaire/QuestionnaireExtraction.kt

```
@@ -11,52 +11,66 @@ object QuestionnaireExtraction {
+                treatmentHistoryCurrentTumor = toList(value(lines, mapping[QuestionnaireKey.TREATMENT_HISTORY_CURRENT_TUMOR])),
+                otherOncologicalHistory = toList(value(lines, mapping[QuestionnaireKey.OTHER_ONCOLOGICAL_HISTORY])),
+                secondaryPrimaries = secondaryPrimaries(lines, mapping[QuestionnaireKey.SECONDARY_PRIMARY]),
+                nonOncologicalHistory = toList(value(lines, mapping[QuestionnaireKey.NON_ONCOLOGICAL_HISTORY])),
+                ihcTestResults = value(lines, mapping[QuestionnaireKey.IHC_TEST_RESULTS])?.let { listOf(trimWhitespaceAndCommas(it)) },
+                pdl1TestResults = toList(value(lines, mapping[QuestionnaireKey.PDL1_TEST_RESULTS])),
+                hasMeasurableDisease = hasMeasurableDisease.curated,
+                hasBrainLesions = brainLesionData.curated?.present(),
+                hasActiveBrainLesions = brainLesionData.curated?.active(),
+                hasCnsLesions = cnsLesionData.curated?.present(),
+                hasActiveCnsLesions = cnsLesionData.curated?.active(),
+                hasBoneLesions = hasBoneLesions.curated,
+                hasLiverLesions = hasLiverLesions.curated,
+                otherLesions = otherLesions(entry, lines, mapping),
+                whoStatus = whoStatus.curated,
+                unresolvedToxicities = toList(value(lines, mapping[QuestionnaireKey.UNRESOLVED_TOXICITIES])),
+                infectionStatus = infectionStatus.curated,
+                ecg = ecg.curated,
+                complications = toList(value(lines, mapping[QuestionnaireKey.COMPLICATIONS])),
+            )
+            val errors = listOf(hasBoneLesions, hasLiverLesions, hasMeasurableDisease, whoStatus, infectionStatus, ecg, stage).flatMap(
+                ValidatedQuestionnaireCuration<*>::errors
+            )
+
+            when {
+                questionnaire.isEmpty() && errors.isEmpty() -> null
+                questionnaire.isEmpty() -> null to errors
```

is "null to errors " really a valid response here?  If a questionnaire is empty, the errors would be about missing data. This will return a pair with the 1st element null, that will be propagated further in the flow, and instead we should take the next questionnaire that is not empty.

---

@andreia-hartwig isn't it better to propagate errors to the user instead of silently skipping them? Otherwise, they could add a new questionnaire with new data and some errors and we would never see it. This was my basis for requesting this change: https://github.com/hartwigmedical/actin/pull/1054#discussion_r2021269684

---

I'm assuming that if Q2 is truly empty, then `errors` should be empty as well and we'll return null: `questionnaire.isEmpty() && errors.isEmpty() -> null`. Otherwise, we would see extraction errors for all empty questionnaires. Often, an extraction error results in a null extracted value, so I worry we'd hide errors if we returned just null for cases where the questionnaire is empty and the errors are not.

---

@jbartletthmf If errors are always empty when the questionnaire is truly empty then there is no issue. However, since questionnaire.isEmpty() returns true when the questionnaire is truly empty, doesn't that mean it will never really return in the second condition?

---

Looking at https://github.com/hartwigmedical/actin/blob/87b6dc56d98cf044090f3ab7c92b7ef3683231da/clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/emc/questionnaire/QuestionnaireCuration.kt#L38-L45
and similar functions, we seem to return `Pair(null, errors)` for many cases where there's an extraction error. I think it's possible to produce an empty `Questionnaire` with non-empty errors, though it may be unlikely.

---

@jbartletthmf This is actually an interesting point, in this case, having an empty questionnaire doesn't actually mean the questionnaire was empty in the feed, it just means none of the validations passed and so all values are null. I wonder if we need to distinguish between those to scenarios - empty in feed vs empty due to failed validations. 

Nevertheless, even in this case, shouldn't we try to see if the next questionnaire on the list of Questionnaire Entries is valid?

---

Going by the [principle of least surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment), I think a new questionnaire entry with errors should never be skipped over without warning. Otherwise, how would we know that we received one?